### DISCUSSION TITLE
Using a sequence of 3D augmentations in AugmentationSequential

### DISCUSSION THREAD

**Discussion by Kaushalya** [kornia/kornia#1345] [Questions] [answered]
### Describe the bug

I'm using Kornia 3D augmentations in a 3D image segmentation project. I found out that AugmentatonSequence doesn't work for 3D transforms.

Input: A 5D torch tensor made of a batch of 3D volumes in the shape of [B X  C X D X H X W]

Example code to reproduce the bug:
```
import kornia.augmentation as K
from kornia.constants import DataKey

transforms = [
        K.RandomHorizontalFlip3D(p=0.5),
        K.RandomVerticalFlip3D(p=0.5),
        K.RandomRotation3D((30, 10, 10), p=0.5)
    ]
aug_list = K.AugmentationSequential(*transforms, data_keys=[DataKey.INPUT, DataKey.MASK])

img_shape = (1, 1, 4, 16, 16)
img = torch.zeros(img_shape)
mask = torch.zeros(img_shape)
aug_list(img, mask)
```
## Error
1. Fails with the following error. `RuntimeError: input shape expected to be in (2, 4) while got torch.Size([1, 1, 4, 16, 16])`
This is caused by the tensor dimension check `_, out_shape = self.autofill_dim(inp, dim_range=(2, 4))` in the AugmentationSequence [[Line 247](https://github.com/kornia/kornia/blob/9ce2bca334fa599fdcfc6abb07b7e9f8533b7d67/kornia/augmentation/container/augment.py#L247)]
2. Even if I give a single 3D image and its mask as two 4D tensors, the augmentations are not applied to the mask because AugmentationBase3D is not a descendant of GeometricAugmentation2D.

## Questions
* Is there a different way to apply a sequence of 3D augmentations on a batch of 3D images?
* Do we need to have a separate base class for 3D geometric augmentations?

Example:
```
# TODO check for 3D geometric augmentations
elif isinstance(module, GeometricAugmentationBase3D) and dcate==DataKey.MASK:
     input, label = ApplyInverse.apply_by_key(input, label, module, param, dcate)
```

### Reproduction steps

```bash
1. Run the code sample I have mentioned in the above section.
```

### Expected behavior

The AugmentationSequence needs to be applied to the batch of images and masks (5D tensors) successfully.

### Environment

- PyTorch Version (e.g., 3.8): 1.8
- OS (e.g., Linux): Linux
- How you installed PyTorch: `pip`
- Build command you used (if compiling from source):
- Python version: 3.8
- CUDA/cuDNN version: 11.4
- GPU models and configuration:
```

**Comment by edgarriba** [comment, marked as answer]
@Kaushalya I think at this moment `AugmentationSequential` is meant for 2d images. However, you can easily create your costumised version following the link below

https://kornia.readthedocs.io/en/latest/applications/image_augmentations.html#customization

```python
import kornia.augmentation as K

class MyAugmentationPipeline(nn.Module):
   def __init__(self) -> None:
      super(MyAugmentationPipeline, self).__init__()
      self.aff = K.RandomAffine(360)
      self.jit = K.ColorJitter(0.2, 0.3, 0.2, 0.3)

   def forward(self, input, mask):
      assert input.shape == mask.shape,
         f"Input shape should be consistent with mask shape, "
         f"while got {input.shape}, {mask.shape}"

      aff_params = self.aff.forward_parameters(input.shape)
      input = self.aff(input, aff_params)
      mask = self.aff(mask, aff_params)

      jit_params = self.jit.forward_parameters(input.shape)
      input = self.jit(input, jit_params)
      mask = self.jit(mask, jit_params)
      return input, mask
```

**Reply by YerePhy** [reply]
Hi, ¿`mask` and `inputs` must have the same shape? I mean if `mask` have different number of channels `RandomAffine` beahaves as expected?  

And just one more question, can I apply different resampling methods for `mask` and `input`?

**Reply by edgarriba** [reply]
in theory if any extra logic happens inside, the warp affine is agnostic to the number of channels. /cc @shijianjian

**Comment by shijianjian** [comment]
For your questions:
> - Is there a different way to apply a sequence of 3D augmentations on a batch of 3D images?

Currently, you may have to use the similar logic as @edgarriba suggested.

> - Do we need to have a separate base class for 3D geometric augmentations?

Yes, adding the support of 3D augmentations require lots of refactoring in 3D augmentation module. You may give it a shot. Eventually, I think we might have ```AugmentationSequential3D``` to do such things.

**Comment by Kaushalya** [comment]
@edgarriba Thanks for the suggestion. I'll proceed with a custom augmentation pipeline.

@shijianjian Thanks for the explanations. I can give it a try at refactoring the 3D augmentation module. 
Shall I go ahead with a PR when I have a working implementation of `AugmentationSequential3D` and refactored 3D augmentations?

**Comment by shijianjian** [comment]
> @edgarriba Thanks for the suggestion. I'll proceed with a custom augmentation pipeline.
> 
> @shijianjian Thanks for the explanations. I can give it a try at refactoring the 3D augmentation module. Shall I go ahead with a PR when I have a working implementation of `AugmentationSequential3D` and refactored 3D augmentations?

Probably separate it into two PRs? The first one will do the refactoring, then the next one can implement the ```AugmentationSequential3D ```.
